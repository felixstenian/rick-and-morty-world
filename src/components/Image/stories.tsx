import { Meta, Story } from '@storybook/react'

import { image_portal_and_brand } from '../../../public/assets/'

import Image from '.'

export default {
  title: 'Image',
  component: Image,
  args: {}
} as Meta

export const Default: Story = (args) => <Image {...args} />

Default.args = {
  src: image_portal_and_brand,
  width: '35%',
  height: '35%',
  alt: 'Rick and morty out portal'
}
