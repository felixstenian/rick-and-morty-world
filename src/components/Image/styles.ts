import styled, { css } from 'styled-components'
import { space, border, layout } from 'styled-system'

import { ImageProps } from '.'

export const Figure = styled.figure<ImageProps>`
  ${({ width, height }) => css`
    width: ${width};
    height: ${height};
  `}
  ${space}
  ${layout}
  ${border}
  overflow: hidden;
`

export const Img = styled.img<{ objectFit?: string } & ImageProps>`
  object-fit: ${({ objectFit }) => objectFit};
  width: 100%;
  height: 100%;
`
