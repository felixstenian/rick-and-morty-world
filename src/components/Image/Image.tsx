import { SpaceProps, BorderProps, LayoutProps } from 'styled-system'

import * as Style from './styles'

export interface ImageProps extends SpaceProps, BorderProps, LayoutProps {
  src?: string
  alt?: string
  width?: string
  height?: string
  objectFit?: string
}

const Image = ({
  src,
  alt,
  width,
  height,
  objectFit,
  ...props
}: ImageProps) => (
  <Style.Figure width={width} height={height} {...props}>
    <Style.Img src={src} alt={alt} objectFit={objectFit} />
  </Style.Figure>
)

export default Image
