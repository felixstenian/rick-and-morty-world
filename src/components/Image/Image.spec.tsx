import { screen } from '@testing-library/react'

import { renderWithTheme } from 'utils'

import Image from '.'

describe('<Image />', () => {
  it('1. should render the reading', () => {
    const src = 'https://t2.tudocdn.net/633314?w=1920'
    const { container } = renderWithTheme(
      <Image src={src} data-testid="image" />
    )

    expect(screen.getByTestId('image')).toBeInTheDocument()
    expect(screen.getByRole('figure')).toBeInTheDocument()
    expect(screen.getByRole('img')).toHaveAttribute('src', src)

    expect(container.firstChild).toMatchSnapshot()
  })
  it('2. should render an image with the given props', () => {
    const src = 'image.jpg'
    const alt = 'alt text'
    const width = '100px'
    const height = '200px'
    const objectFit = 'cover'

    renderWithTheme(
      <Image
        src={src}
        alt={alt}
        width={width}
        height={height}
        objectFit={objectFit}
        data-testid="image"
      />
    )

    expect(screen.getByRole('figure')).toHaveAttribute('width', width)
    expect(screen.getByRole('figure')).toHaveAttribute('height', height)

    expect(screen.getByRole('img')).toHaveAttribute('src', src)
    expect(screen.getByRole('img')).toHaveAttribute('alt', alt)
    expect(screen.getByRole('img')).toHaveStyle({ objectFit: objectFit })
  })
})
