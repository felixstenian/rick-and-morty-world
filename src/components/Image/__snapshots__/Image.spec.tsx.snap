// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> 1. should render the reading 1`] = `
.c0 {
  overflow: hidden;
}

.c1 {
  width: 100%;
  height: 100%;
}

<figure
  class="c0"
  data-testid="image"
>
  <img
    class="c1"
    src="https://t2.tudocdn.net/633314?w=1920"
  />
</figure>
`;
